src/mon/AuthMonitor.cc

```diff
@@ -1012,6 +1015,32 @@ int AuthMonitor::do_osd_new(
   return 0;
 }
 
+bool AuthMonitor::valid_caps(const vector<string>& caps, ostream *out)
+{
+  for (vector<string>::const_iterator p = caps.begin();
+       p != caps.end(); p += 2) {
+    if (!p->empty() && (*p == "mon" || *p == "mgr")) {
+      MonCap tmp;
+      if (!tmp.parse(*(p+1), out)) {
+	return false;
+      }
+    }
+    if (!p->empty() && *p == "osd") {
+      OSDCap ocap;
+      if (!ocap.parse(*(p+1), out)) {
+	return false;
+      }
+    }
+    if (!p->empty() && *p == "mds") {
+      MDSAuthCaps mdscap;
+      if (!mdscap.parse(g_ceph_context, *(p+1), out)) {
+	return false;
+      }
+    }
+  }
+  return true;
```

Comment:
Written by gregsfortytwo
>This construction appears to validate the caps (by falling through all the test cases) if it doesn't match any of the known cap users. That seems bad?

Written by bellaalleb
>Here it only try to validate the caps when it's type is either "mon" or "osd" or "mds" and won't validate any other unknown types.
Do you mean that it's bad for unknown types?

Written by gregsfortytwo
>Yeah, I don't think we want to accept capabilities that don't match a known type. Otherwise somebody could typo and end up with caps on the "min" and none on the mon, or similar.

Reply:
Thanks, cleaned up the checks and error messages to be more informative/robust too.